### PR TITLE
Add VolumeAttachment event registration to CSI volume limits plugin

### LINKED ISSUE #128324: [Failing Test] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode / default fs)] volume-stress multiple pods should access different volumes repeatedly [Slow] [Serial]
### Which jobs are failing?

- sig-release-master-informing
  - gce-cos-master-serial

### Which tests are failing?

- `Kubernetes e2e suite.[It] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] volume-stress multiple pods should access different volumes repeatedly [Slow] [Serial]`
- `Kubernetes e2e suite: [It] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] volume-stress multiple pods should access different volumes repeatedly [Slow] [Serial]`

### Since when has it been failing?

2024-10-23 12:40 PDT

[352056f09](https://github.com/search?q=repo:kubernetes/kubernetes+352056f09&type=commits)

### Testgrid link

https://testgrid.k8s.io/sig-release-master-informing#gce-cos-master-serial

### Reason for failure (if possible)

```
{ failed [FAILED] Failed to wait for pod-6 [&Pod{ObjectMeta:{pod-9420b3ba-fa6f-4da2-bab3-b57bc29584fb  stress-379    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},Spec:PodSpec{Volumes:[]Volume{Volume{Name:volume1,VolumeSource:VolumeSource{HostPath:nil,EmptyDir:nil,GCEPersistentDisk:nil,AWSElasticBlockStore:nil,GitRepo:nil,Secret:nil,NFS:nil,ISCSI:nil,Glusterfs:nil,PersistentVolumeClaim:&PersistentVolumeClaimVolumeSource{ClaimName:csi-hostpathdc6bb,ReadOnly:false,},RBD:nil,FlexVolume:nil,Cinder:nil,CephFS:nil,Flocker:nil,DownwardAPI:nil,FC:nil,AzureFile:nil,ConfigMap:nil,VsphereVolume:nil,Quobyte:nil,AzureDisk:nil,PhotonPersistentDisk:nil,PortworxVolume:nil,ScaleIO:nil,Projected:nil,StorageOS:nil,CSI:nil,Ephemeral:nil,Image:nil,},},},Containers:[]Container{Container{Name:write-pod,Image:registry.k8s.io/e2e-test-images/busybox:1.36.1-1,Command:[/bin/sh -c trap exit TERM; while true; do sleep 1; done],Args:[],WorkingDir:,Ports:[]ContainerPort{},Env:[]EnvVar{},Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},Claims:[]ResourceClaim{},},VolumeMounts:[]VolumeMount{VolumeMount{Name:volume1,ReadOnly:false,MountPath:/mnt/volume1,SubPath:,MountPropagation:nil,SubPathExpr:,RecursiveReadOnly:nil,},},LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:,ImagePullPolicy:,SecurityContext:&SecurityContext{Capabilities:nil,Privileged:*false,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,RunAsGroup:nil,ProcMount:nil,WindowsOptions:nil,SeccompProfile:nil,AppArmorProfile:nil,},Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[]EnvFromSource{},TerminationMessagePolicy:,VolumeDevices:[]VolumeDevice{},StartupProbe:nil,ResizePolicy:[]ContainerResizePolicy{},RestartPolicy:nil,},},RestartPolicy:OnFailure,TerminationGracePeriodSeconds:nil,ActiveDeadlineSeconds:nil,DNSPolicy:,NodeSelector:map[string]string{},ServiceAccountName:,DeprecatedServiceAccount:,NodeName:,HostNetwork:false,HostPID:false,HostIPC:false,SecurityContext:&PodSecurityContext{SELinuxOptions:&SELinuxOptions{User:,Role:,Type:,Level:s0:c0,c1,},RunAsUser:nil,RunAsNonRoot:nil,SupplementalGroups:[],FSGroup:*1000,RunAsGroup:nil,Sysctls:[]Sysctl{},WindowsOptions:nil,FSGroupChangePolicy:nil,SeccompProfile:nil,AppArmorProfile:nil,SupplementalGroupsPolicy:nil,},ImagePullSecrets:[]LocalObjectReference{},Hostname:,Subdomain:,Affinity:nil,SchedulerName:,InitContainers:[]Container{},AutomountServiceAccountToken:nil,Tolerations:[]Toleration{},HostAliases:[]HostAlias{},PriorityClassName:,Priority:nil,DNSConfig:nil,ShareProcessNamespace:nil,ReadinessGates:[]PodReadinessGate{},RuntimeClassName:nil,EnableServiceLinks:nil,PreemptionPolicy:nil,Overhead:ResourceList{},TopologySpreadConstraints:[]TopologySpreadConstraint{},EphemeralContainers:[]EphemeralContainer{},SetHostnameAsFQDN:nil,OS:nil,HostUsers:nil,SchedulingGates:[]PodSchedulingGate{},ResourceClaims:[]PodResourceClaim{},},Status:PodStatus{Phase:,Conditions:[]PodCondition{},Message:,Reason:,HostIP:,PodIP:,StartTime:<nil>,ContainerStatuses:[]ContainerStatus{},QOSClass:,InitContainerStatuses:[]ContainerStatus{},NominatedNodeName:,PodIPs:[]PodIP{},EphemeralContainerStatuses:[]ContainerStatus{},Resize:,ResourceClaimStatuses:[]PodResourceClaimStatus{},HostIPs:[]HostIP{},},}] turn into running status. Error: Timed out after 300.001s.
Expected Pod to be in <v1.PodPhase>: "Running"
Got instead:
    <*v1.Pod | 0xc002a26908>: 
        metadata:
          creationTimestamp: "2024-10-23T20:24:51Z"
          managedFields:
......(truncate)
```

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig storage
@kubernetes/release-team-release-signal

### PR DESCRIPTION
**Pull request by torredil**
#### What type of PR is this?

/kind failing-test

#### What this PR does / why we need it:

See https://github.com/kubernetes/kubernetes/issues/128324#issuecomment-2436714741 for context.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/128324

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by torredil** [comment]
cc: @carlory @sanposhiho

File: pkg/scheduler/framework/plugins/nodevolumelimits/csi.go [hunk 1]
```diff
@@ -89,6 +89,7 @@ func (pl *CSILimits) EventsToRegister(_ context.Context) ([]framework.ClusterEve
 		{Event: framework.ClusterEvent{Resource: framework.CSINode, ActionType: framework.Add}},
 		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Delete}, QueueingHintFn: pl.isSchedulableAfterPodDeleted},
 		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterPVCAdded},
+		{Event: framework.ClusterEvent{Resource: framework.VolumeAttachment, ActionType: framework.Delete}},
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:92]: Can you open an issue for QHint implementation for this event so that we don't forget?

**Review by sanposhiho — APPROVED** [review]
/approve

Leave `/lgtm` to @carlory or someone else

File: pkg/scheduler/framework/plugins/nodevolumelimits/csi.go (hunk 1 above)
Comment (torredil) [reply pkg/scheduler/framework/plugins/nodevolumelimits/csi.go:92]: https://github.com/kubernetes/kubernetes/issues/128347

**Comment by carlory** [comment]
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/eventhandlers.go#L341 and https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/types.go#L210 should be updated as well to match the new resource type.

**Comment by sanposhiho** [comment]
Regarding eventhandlers.go, it's done already, right?
And, types.go, Yes, good point, I missed that part. We need it otherwise scheduler-perf doesn't measure the metrics.

/remove-approve

**Comment by carlory** [comment]
> Regarding eventhandlers.go, it's done already, right?

Yes.

**Comment by torredil** [comment]
Updated https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/types.go#L210 in latest rev.

cc: @carlory @sanposhiho

**Review by sanposhiho — APPROVED** [review]
/approve

**Comment by carlory** [comment]
Thanks！
/lgtm

**Pull request merged**